components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dao/JDBCSubscriberDAOImpl.java

```diff
@@ -0,0 +1,372 @@
+                        updateSubscriberPropertiesPrepStmt.setString(SUBSCRIBER_ID, subscriberId);
+                        updateSubscriberPropertiesPrepStmt.setInt(TENANT_ID, tenantId);
+                        updateSubscriberPropertiesPrepStmt.addBatch();
+                    }
+                }
+            }
+            updateSubscriberPropertiesPrepStmt.executeBatch();
+            IdentityDatabaseUtil.commitTransaction(connection);
+        } catch (SQLException e) {
+            IdentityDatabaseUtil.rollbackTransaction(connection);
+            throw new EntitlementException("Error while updating subscriber details", e);
+        } finally {
+            IdentityDatabaseUtil.closeConnection(connection);
+        }
+    }
+
+    /**
+     * Sets the base64 encoded secret value of the secret subscriber properties, if it has been updated.
+     *
+     * @param holder    publisher data holder
+     * @param oldHolder existing publisher data holder
+     */
+    private void populateProperties(PublisherDataHolder holder, PublisherDataHolder oldHolder) {
+
+        PublisherPropertyDTO[] propertyDTOs = holder.getPropertyDTOs();
+
+        for (PublisherPropertyDTO dto : propertyDTOs) {
+            if (StringUtils.isNotBlank(dto.getId()) && StringUtils.isNotBlank(dto.getValue()) && dto.isSecret()) {
+                PublisherPropertyDTO propertyDTO = null;
+                if (oldHolder != null) {
+                    propertyDTO = oldHolder.getPropertyDTO(dto.getId());
+                }
+                if (propertyDTO == null || !propertyDTO.getValue().equalsIgnoreCase(dto.getValue())) {
+                    try {
+                        String encryptedValue =
+                                CryptoUtil.getDefaultCryptoUtil().encryptAndBase64Encode(dto.getValue().getBytes());
+                        dto.setValue(encryptedValue);
+                    } catch (CryptoException e) {
+                        LOG.error("Error while encrypting secret value of subscriber. Secret would not be persisted.",
+                                e);
+                    }
+                }
+            }
+        }
+    }
+
+    private PublisherDataHolder getPublisherDataHolder(ResultSet resultSet, boolean returnSecrets) throws SQLException {
+
+        List<PublisherPropertyDTO> propertyDTOList = new ArrayList<>();
+        String moduleName;
+        do {
+            PublisherPropertyDTO dto = new PublisherPropertyDTO();
+
+            dto.setId(resultSet.getString(PROPERTY_ID));
+            dto.setValue(resultSet.getString(PROPERTY_VALUE));
+            dto.setDisplayName(resultSet.getString(DISPLAY_NAME));
+            dto.setDisplayOrder(resultSet.getInt(DISPLAY_ORDER));
+            dto.setRequired(resultSet.getBoolean(IS_REQUIRED));
+            dto.setSecret(resultSet.getBoolean(IS_SECRET));
+            dto.setModule(resultSet.getString(MODULE));
+
+            if (dto.isSecret() && returnSecrets) {
```

Comment:
with current logic, its always returns the secret either in encrypted form or decrypted form? Is its the expected behavior? if so, lets rename `returnSecrets` to something like `returnDecryptedSecrets`.

Reply:
> with current logic, its always returns the secret either in encrypted form or decrypted form? Is its the expected behavior?

if not, `dto.setValue(resultSet.getString(PROPERTY_VALUE));` should move inside to this if block.

Reply:
yes `returnSecrets` -> true returns the decrypted secrets, and false returns encrypted secrets. so will rename